text/projects/Maciej_Nems/Simple_String_Matching_k_Mistakes.tex

```diff
@@ -0,0 +1,126 @@
+\usepackage{polski}
+\usepackage[utf8]{inputenc}
+\usepackage[polish]{babel}
+\usepackage[]{amsthm} %lets us use \begin{proof}
+\usepackage[]{amssymb} %gives us the character \varnothing
+\usepackage{amsmath}
+\usepackage{algorithm}
+\usepackage{algorithmic}
+\newcommand{\algorithmicinput}{\textbf{input}}
+\newcommand{\INPUT}{\item[\algorithmicinput]}
+
+\title{Simple string matching with k mismatches \\ \large W oparciu o ,,Simple and efficient string matching with k mismatches"\\ R. Grossi, F. Luccio}
+\author{Maciej Nemś}
+\date{Maj 2022}
+
+\begin{document}
+
+\maketitle
+
+\section{Wstęp}
+Problem \textit{string matching with k mismatches} (SMK) polega na tym, że dla tekstu $T$ oraz dla wzorca $W$ takich, że $|T| = n$, $|W| = m$ znajdujemy wszystkie wystąpienia $W$ w $T$ z co najwyżej $k < m$ różnymi symbolami. W pracy autorzy przedstawiają dwa algorytmy rozwiązujące ten problem. Oba opierają się na liczbie permutacji znaków $W$ w $T$ z co najwyżej $k$ błędami. Pierwszy algorytm działa w czasie $O(n\log|A_W|+rm)$, gdzie $r$ to liczba wystąpień permutacji $W$ w $T$ z $k$ błędami, natomiast $A_w$, to zbiór znaków w $W$. Drugi algorytm działa w czasie $O(n\log|A_W| +dk)$, gdzie $d \leq r$, to liczba różnych wystąpień permutacji $W$ w $T$ z $k$ błędami. Oba algorytmy najpierw szukają wszystkich miejsc, gdzie w $T$ jest jakaś permutacja $W$ z $k$ błędami, a następnie zwracają z tych miejsc wszystkie te, które mają co najwyżej $k$ błędów.
+
+\section{Opis algorytmów}
+
+\subsection{Szukanie permutacji}
+\subsubsection*{Algorytm}
+Oba algorytmy rozwiązujące problem SMK wymagają najpierw obliczenia wszystkich miejsc w tekście $T$, gdzie zaczyna się jakaś permutacja wzorca $W$ z co najwyżej $k$ błędami. Do szukania permutacji będziemy potrzebować kolejki FIFO $Q$, która będzie trzymać symbole z $T$ oraz struktury $C$, która będzie trzymała liczniki dla każdego znaku występującego w $A_T$. Licznik $Z$ będzie zliczał liczbę błędów.
+\par
+Algorytm dla każdego znaku w tekście najpierw go dorzuca do kolejki i dekrementuje licznik. Jeśli $C[znak] < 0$, to znaczy, że wykorzystaliśmy już wszystkie wystąpienia tego znaku w $W$, więc mamy błąd (inkrementujemy $K$). Teraz tak długo, aż błędów w kolejce jest ponad $k$, usuwamy znaki z kolejki i odpowiednio dekrementujemy $K$. Po pętli, jeśli rozmiar $Q$ równa się $m$, to zebraliśmy tyle znaków, co długość wzorca, mamy co najwyżej $K$ błędów, czyli wiemy, że to jest permutacja wzorca z co najwyżej $K$ błędami i możemy ją zgłosić jako match. Dodatkowo, aby zawsze na początku pętli rozmiar $Q$ wynosił co najwyżej $m-1$, to usuwamy pierwszy element.
+\begin{algorithm}[H]
+\caption{Permutation Matching}
+\begin{algorithmic} 
+\INPUT{$T$, $W$, $n$, $m$, $k$}
+\IF{$n < m$}
+    \STATE{end with no matches}
+\ENDIF
+\STATE{\textbf{Preprocessing:} For each $a \in A_W$ set $C[a]$ to count of $a$ in $W$; Substitute all $t \in T$, $t \notin A_T$ with special symbol $\# \notin A_P$ to get $\tilde{T}$; Set $C[\#] = 0$}
```

Comment:
1. It was my mistake. I wrote $P$ instead of $W$, as I was used to use $P$ as in Pattern instead $W$ as in Wzorzec?
2. For all $t \in T$ if $t \notin A_W$, substitute it with wymbol $\\%\notin A_W$ to get new word $\tilde{T}$.